components/file_system/src/rate_limiter.rs

```diff
@@ -1,43 +1,363 @@
+                        *last_refill_time = now;
+                        break self.refill_and_request(bytes);
+                    }
+                }
+            }
+        }
+    }
+
+    pub async fn async_request(&self, bytes: usize, priority: IOPriority) -> usize {
+        let cached_bytes_per_refill = self.bytes_per_refill.load(Ordering::Relaxed);
+        if cached_bytes_per_refill == 0 {
+            return bytes;
+        }
+        if priority == IOPriority::High {
+            self.consumed.fetch_add(bytes, Ordering::Relaxed);
+            return bytes;
+        }
+        loop {
+            if let Some(bytes) = self.request_fast(cached_bytes_per_refill, bytes) {
+                break bytes;
+            }
+            let mut last_refill_time = self.last_refill_time.lock().unwrap();
+            if self.consumed.load(Ordering::Relaxed) < cached_bytes_per_refill {
+                continue;
+            }
+            if let Some(calibrator) = &self.calibrator {
+                let calibrated = calibrator.calibrate();
+                self.consumed.store(calibrated, Ordering::Relaxed);
+                if calibrated < cached_bytes_per_refill {
+                    continue;
+                }
+            }
+            let now = Instant::now();
+            if now > *last_refill_time {
+                let since_last_refill = now.duration_since(*last_refill_time);
+                if since_last_refill >= self.refill_period {
+                    *last_refill_time = now;
+                    break self.refill_and_request(bytes);
+                } else {
+                    let cached_last_refill_time = *last_refill_time;
+                    let (mut last_refill_time, timed_out) = self
+                        .condv
+                        .async_wait_timeout(
+                            &self.last_refill_time,
+                            last_refill_time,
+                            self.refill_period - since_last_refill,
+                        )
+                        .await;
+                    let now = Instant::now();
+                    if timed_out && *last_refill_time == cached_last_refill_time {
+                        *last_refill_time = now;
+                        break self.refill_and_request(bytes);
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl Default for PerTypeIORateLimiter {
+    fn default() -> PerTypeIORateLimiter {
+        PerTypeIORateLimiter::new(0, Duration::from_millis(10))
+    }
+}
 
-/// No-op limiter
 /// An instance of `IORateLimiter` should be safely shared between threads.
 #[derive(Debug)]
 pub struct IORateLimiter {
-    refill_bytes: atomic::AtomicUsize,
+    // IOType::Other slot is used to store type-less limiter
+    write_limiters: [PerTypeIORateLimiter; IOType::VARIANT_COUNT],
+    read_limiters: [PerTypeIORateLimiter; IOType::VARIANT_COUNT],
+    total_limiters: [PerTypeIORateLimiter; IOType::VARIANT_COUNT],
```

Comment:
Is there any case for limiting read+write IO? I think we just need to limit write/read separately?

Reply:
I think cloud disk for instance requires to limit total throughput.